plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -354,6 +357,55 @@ public boolean isSourceModified() {
         return sourceModified;
     }
 
+    static Object deepCopy(Object value) {
+        if (value instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<Object, Object> mapValue = (Map<Object, Object>) value;
+            return deepCopyMap(mapValue);
+        } else if (value instanceof List) {
+            @SuppressWarnings("unchecked")
+            List<Object> listValue = (List<Object>) value;
+            return deepCopyList(listValue);
+        } else {
+            return value;
```

Comment:
we don't clone objects because we will have a limited set of types to handle and besides and lists and maps we shouldn't need to worry about references right?